# Project 3 -- Lukas Anell

In [1]:
# /anvil/projects/tdm/data/grouping/vgchartz-2024.csv
# /anvil/projects/tdm/data/grouping/googleplaystore.csv

## Question 1

In [38]:
import pandas as pd
import numpy as np


video_games = pd.read_csv("/anvil/projects/tdm/data/grouping/vgchartz-2024.csv")
video_games.head()

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update
0,/games/boxart/full_6510540AmericaFrontccc.jpg,Grand Theft Auto V,PS3,Action,Rockstar Games,Rockstar North,9.4,20.32,6.37,0.99,9.85,3.12,2013-09-17,NaN
1,/games/boxart/full_5563178AmericaFrontccc.jpg,Grand Theft Auto V,PS4,Action,Rockstar Games,Rockstar North,9.7,19.39,6.06,0.60,9.71,3.02,2014-11-18,2018-01-03
2,/games/boxart/827563ccc.jpg,Grand Theft Auto: Vice City,PS2,Action,Rockstar Games,Rockstar North,9.6,16.15,8.41,0.47,5.49,1.78,2002-10-28,NaN
3,/games/boxart/full_9218923AmericaFrontccc.jpg,Grand Theft Auto V,X360,Action,Rockstar Games,Rockstar North,NaN,15.86,9.06,0.06,5.33,1.42,2013-09-17,NaN
4,/games/boxart/full_4990510AmericaFrontccc.jpg,Call of Duty: Black Ops 3,PS4,Shooter,Activision,Treyarch,8.1,15.09,6.18,0.41,6.05,2.44,2015-11-06,2018-01-14


In [39]:
video_games.groupby("genre", as_index=False)["critic_score"].mean().sort_values(by="critic_score", ascending=False).head()

,genre,critic_score
14,Sandbox,9.200000
19,Visual Novel,8.500000
8,Music,7.992593
6,MMO,7.781818
4,Education,7.750000


In [40]:
video_games.groupby("console", as_index=False)["total_sales"].sum().sort_values(by="total_sales", ascending=False).head()

,console,total_sales
53,PS2,1027.76
73,X360,859.79
54,PS3,839.70
52,PS,546.25
55,PS4,539.92


I had to look up how to sort the values, but it wasn't too difficult in the end.

1.1: The Sandbox genre had the highest average critic_score.

1.2: The PS2 had the highest number of total_sales.

1.3: Grouping data in Python is useful because it allows you to see how certain parts of the data are related to each other. For example, we can group the data to find which game genre has the highest critic score.

## Question 2

In [53]:
video_games["high_rated"] = video_games["critic_score"] >= 7
video_games["high_rated"].value_counts()

high_rated
False    59689
True      4327
Name: count, dtype: int64

In [55]:
video_games["low_rated"] = video_games["critic_score"] < 7
video_games["low_rated"].value_counts()

low_rated
False    61665
True      2351
Name: count, dtype: int64

In [56]:
video_games['critic_score'].isna().sum()

np.int64(57338)

In [57]:
sports_hits = video_games[(video_games["genre"] == "Sports") & (video_games["high_rated"])].copy()
sports_hits.head()

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
15,/games/boxart/full_5257064AmericaFrontccc.jpg,FIFA 18,PS4,Sports,EA Sports,EA Vancouver,8.3,11.80,1.27,0.15,8.64,1.73,2017-09-29,2018-04-02,True,False
17,/games/boxart/full_7661370AmericaFrontccc.jpg,FIFA 17,PS4,Sports,Electronic Arts,EA Canada,8.9,10.94,1.26,0.12,7.95,1.61,2016-09-27,2018-03-28,True,False
31,/games/boxart/full_862315AmericaFrontccc.jpg,FIFA 16,PS4,Sports,EA Sports,EA Canada,8.5,8.22,1.15,0.07,5.77,1.23,2015-09-22,2018-04-01,True,False
48,/games/boxart/full_fifa-soccer-12_475AmericaFr...,FIFA Soccer 12,PS3,Sports,EA Sports,EA Canada,9.2,6.65,0.84,0.11,4.33,1.37,2011-09-27,NaN,True,False
55,/games/boxart/full_189344AmericaFrontccc.jpg,FIFA 15,PS4,Sports,EA Sports,EA Canada,8.1,6.32,0.83,0.05,4.49,0.94,2014-09-23,2018-03-23,True,False


In [58]:
sports_hits[sports_hits['critic_score'] == 9.5]

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
75,/games/boxart/1473680ccc.jpg,Madden NFL 2004,PS2,Sports,EA Sports,EA Tiburon,9.5,5.23,4.26,0.01,0.26,0.71,2003-08-12,NaN,True,False
114,/games/boxart/4564394ccc.jpg,Madden NFL 2005,PS2,Sports,EA Sports,EA Tiburon,9.5,4.53,4.18,0.01,0.26,0.08,2004-08-09,NaN,True,False
4359,/games/boxart/8490939ccc.jpg,SSX 3,GC,Sports,EA Sports BIG,EA Canada,9.5,0.38,0.29,NaN,0.08,0.01,2003-10-20,NaN,True,False


In [59]:
sports_fails = video_games[(video_games["genre"] == "Sports") & (video_games["low_rated"])].copy()
sports_fails.head()

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
109,/games/boxart/full_2431298AmericaFrontccc.jpg,FIFA 15,PS3,Sports,EA Sports,EA Canada,6.9,4.56,0.59,0.04,3.25,0.68,2014-09-23,2018-03-23,False,True
110,/games/boxart/full_1173134AmericaFrontccc.jpg,Mario & Sonic at the Olympic Winter Games,Wii,Sports,Sega,Sega,6.8,4.54,1.87,0.22,1.97,0.48,2009-10-13,NaN,False,True
272,/games/boxart/full_7298854PALFrontccc.jpg,FIFA 15,X360,Sports,EA Sports,EA Canada,6.9,2.91,0.80,NaN,1.90,0.21,2014-09-23,2018-03-23,False,True
277,/games/boxart/full_9888617AmericaFrontccc.jpg,Sega Superstars Tennis,X360,Sports,Sega,Sumo Digital,6.9,2.89,1.75,NaN,0.86,0.28,2008-03-18,NaN,False,True
380,/games/boxart/full_pro-evolution-soccer-2011_2...,Pro Evolution Soccer 2011,PS3,Sports,Konami,Konami,6.0,2.41,0.29,0.54,1.18,0.41,2010-10-05,NaN,False,True


In [60]:
sports_fails[sports_fails['critic_score'] == 2.8]

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
8292,/games/boxart/full_554728AmericaFrontccc.jpg,Major League Baseball 2K10,DS,Sports,2K Sports,Visual Concepts,2.8,0.15,0.14,NaN,NaN,0.01,2010-03-02,NaN,False,True
59297,/games/boxart/full_mlb-slam_460AmericaFront.jpg,MLB Slam!,NGage,Sports,THQ,THQ,2.8,NaN,NaN,NaN,NaN,NaN,2008-10-31,NaN,False,True


When searching online, I found [np.where](https://numpy.org/devdocs/reference/generated/numpy.where.html) to make the high and low rated columns, so I didn't run into the NA values issue.
For the deliverable asking about the true/false counts, I think it works to just do .sum() on the value_counts()?
I switched to normal boolean logic because I realized np.where wasn't doing what I wanted.

2.1:
high_score: 4327 True, 59689 False
low_score: 2351 True, 61665 False

2.2: Madden NFL 2004, Madden NFL 2005, SSX 3

2.3: Major League Basketball 2K10, MLB Slam!

## Question 3

In [61]:
video_games.head(5)

,img,title,console,genre,publisher,developer,critic_score,total_sales,na_sales,jp_sales,pal_sales,other_sales,release_date,last_update,high_rated,low_rated
0,/games/boxart/full_6510540AmericaFrontccc.jpg,Grand Theft Auto V,PS3,Action,Rockstar Games,Rockstar North,9.4,20.32,6.37,0.99,9.85,3.12,2013-09-17,NaN,True,False
1,/games/boxart/full_5563178AmericaFrontccc.jpg,Grand Theft Auto V,PS4,Action,Rockstar Games,Rockstar North,9.7,19.39,6.06,0.60,9.71,3.02,2014-11-18,2018-01-03,True,False
2,/games/boxart/827563ccc.jpg,Grand Theft Auto: Vice City,PS2,Action,Rockstar Games,Rockstar North,9.6,16.15,8.41,0.47,5.49,1.78,2002-10-28,NaN,True,False
3,/games/boxart/full_9218923AmericaFrontccc.jpg,Grand Theft Auto V,X360,Action,Rockstar Games,Rockstar North,NaN,15.86,9.06,0.06,5.33,1.42,2013-09-17,NaN,False,False
4,/games/boxart/full_4990510AmericaFrontccc.jpg,Call of Duty: Black Ops 3,PS4,Shooter,Activision,Treyarch,8.1,15.09,6.18,0.41,6.05,2.44,2015-11-06,2018-01-14,True,False


In [83]:
genre_counts = video_games.groupby("genre", as_index=False)["title"].count()
genre_counts = genre_counts.rename(columns={"title": "game_count"})
genre_counts

,genre,game_count
0,Action,8557
1,Action-Adventure,1877
2,Adventure,6260
3,Board Game,33
4,Education,35
5,Fighting,2367
6,MMO,115
7,Misc,9304
8,Music,297
9,Party,151


In [88]:
genre_counts_sorted = genre_counts.sort_values(by="game_count", ascending=False)
genre_counts_sorted.head(10)

,genre,game_count
7,Misc,9304
0,Action,8557
2,Adventure,6260
13,Role-Playing,5721
17,Sports,5586
15,Shooter,5410
10,Platform,4001
18,Strategy,3685
11,Puzzle,3521
12,Racing,3425


In [92]:
console_genre_counts = video_games.groupby(["genre", "console"], as_index=False)["title"].count()
console_genre_counts.head()

,genre,console,title
0,Action,2600,302
1,Action,3DO,12
2,Action,3DS,198
3,Action,5200,26
4,Action,7800,21


I like how easy it is to rename columns.

3.1: The Misc genre has the most games in the dataset.

3.2: See above.

3.3: See above.

## Question 4

In [100]:
app_store = pd.read_csv("/anvil/projects/tdm/data/grouping/googleplaystore.csv")
app_store.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


In [101]:
app_store.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

In [106]:
app_store["Clean_Reviews"] = app_store["Reviews"].str.replace(",", "", regex=False)
app_store = app_store[app_store["Clean_Reviews"] != "3.0M"]
app_store["Numeric_Reviews"] = app_store["Clean_Reviews"].astype(float)
app_store["Numeric_Reviews"].head()

0       159.0
1       967.0
2     87510.0
3    215644.0
4       967.0
Name: Numeric_Reviews, dtype: float64

In [105]:
app_store["rating_per_review"] = app_store["Rating"] / app_store["Numeric_Reviews"]
app_store["rating_per_review"].head()

0    0.025786
1    0.004033
2    0.000054
3    0.000021
4    0.004447
Name: rating_per_review, dtype: float64

I think it's very interesting that you can just divide columns like I did in my code. I guess it's more convenient than having to manually iterate over both columns.

4.1: "3.0M" was causing an error.

4.2: See above.

4.3: See above.

## Question 5

In [115]:
app_store["Content Rating"].unique().shape

(6,)

In [116]:
app_store["Content Rating"].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated'], dtype=object)

In [108]:
app_store["New_Content_Ratings"] = app_store["Content Rating"].str.upper()
app_store["New_Content_Ratings"].head()

0    EVERYONE
1    EVERYONE
2    EVERYONE
3        TEEN
4    EVERYONE
Name: New_Content_Ratings, dtype: object

In [112]:
app_store.groupby("New_Content_Ratings")["App"].count().sort_values(ascending=False)

New_Content_Ratings
EVERYONE           8714
TEEN               1208
MATURE 17+          499
EVERYONE 10+        414
ADULTS ONLY 18+       3
UNRATED               2
Name: App, dtype: int64

This one is pretty straightforward to me. Unique is, of course, very useful, but it's also something we used last semester in R.

5.1: See above.

5.2: See above.

5.3: EVERYONE has the most apps.

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
